apm-agent-plugins/apm-opentelemetry/apm-opentelemetry-plugin/src/main/java/co/elastic/apm/agent/opentelemetry/sdk/OTelSpan.java

```diff
@@ -0,0 +1,279 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package co.elastic.apm.agent.opentelemetry.sdk;
+
+import co.elastic.apm.agent.impl.context.Url;
+import co.elastic.apm.agent.impl.transaction.AbstractSpan;
+import co.elastic.apm.agent.impl.transaction.OTelSpanKind;
+import co.elastic.apm.agent.impl.transaction.Outcome;
+import co.elastic.apm.agent.impl.transaction.Transaction;
+import co.elastic.apm.agent.sdk.logging.Logger;
+import co.elastic.apm.agent.sdk.logging.LoggerFactory;
+import co.elastic.apm.agent.util.LoggerUtils;
+import co.elastic.apm.agent.util.VersionUtils;
+import io.opentelemetry.api.OpenTelemetry;
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.SpanContext;
+import io.opentelemetry.api.trace.StatusCode;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+public class OTelSpan implements Span {
+    private static final Logger eventLogger = LoggerUtils.logOnce(LoggerFactory.getLogger(OTelSpan.class));
+
+    private final AbstractSpan<?> span;
+
+    public OTelSpan(AbstractSpan<?> span) {
+        this.span = span;
+        span.incrementReferences();
```

Comment:
This is an "intentional leak" to have a similar behavior like our Agent Public API where spans/transactions are not recycled.
Given the OTel API provides ways to wrap an active span into `Callable` or `Runnable` there are less opportunities for end-users to mis-use it, so maybe in the future we could optionally enable recycling.

We don't have any way to distinguish OTel API usages, so "custom instrumentation" and "library that explicitly uses it", at best we could make some heuristics by the provided fields and values but that sounds quite brittle. What would be the benefit to see a difference between the two ?

Reply:
It's not only about misusing the API. It's valid for a span to be used as a parent after it has ended. This would set the reference count to 0, yet it can't be recycled as it may be used as a parent of another span.

Reply:
I see. Then maybe we can do something even safer that will move the lifecycle management responsibility from us to the using library through the API, so that "advanced" libraries can optionally tell us when a span can be recycled.
I am not sure about the feasibility of convincing OTel to add an optional `recycle()`/`dispose()` functionality to the API, but we can use the current API, e.g. with a special "recycle" event through `addEvent`, or a second invocation of `end`.
It may be a significant advantage if we can pull off recycling for "specialist" libraries.

Reply:
We thought of a solution that may be as efficient and simple to get done: creating a tiny "OTel API extension" that contains not much more than `Tracer.recycle(Object)` with a noop implementation that can be overridden dynamically with a concrete implementation. Then it's both efficient, robust and doesn't interfere with agent vendor.
But first thing is to measure what the actual overhead of not recycling is.